api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
     tiersForRemoval.forEach(tier -> accessTierService.removeUserFromTier(dbUser, tier));
   }
 
-  public boolean isCompleteAndNotExpired(Timestamp completionTime) {
-    Timestamp expirationTime =
-        new Timestamp(
-            clock.millis()
-                - TimeUnit.MILLISECONDS.convert(
-                    configProvider.get().accessRenewal.expiryDays, TimeUnit.DAYS));
-    if (configProvider.get().access.enableAccessRenewal) {
-      return completionTime != null && expirationTime.before(completionTime);
+  private class ModuleTimes {
+    public Optional<Timestamp> completion;
+    public Optional<Timestamp> bypass;
+
+    public ModuleTimes(Timestamp nullableCompletion, Timestamp nullableBypass) {
+      completion = Optional.ofNullable(nullableCompletion);
+      bypass = Optional.ofNullable(nullableBypass);
+    }
+
+    public boolean isBypassed() {
+      return bypass.isPresent();
+    }
+
+    public boolean isComplete() {
+      return completion.isPresent();
+    }
+
+    public Optional<Timestamp> getExpiration() {
```

Comment:
we can shrink this a bit by reordering and using Optional.map().  Here is my thought:

if (isBypassed() || !config.enableRenewal) return empty;

if (expiryDays == null) throw ServerError/500 or something

return completion.map(c -> new Timestamp(c.getTime() + expiry);

Reply:
Very nice!